base/initdefs.jl

```diff
@@ -354,6 +354,7 @@ const atexit_hooks = Callable[
     () -> Filesystem.temp_cleanup_purge(force=true)
 ]
 const _atexit_hooks_lock = ReentrantLock()
+global _atexit_hooks_finished::Bool = false
```

Comment:
I'm thinking this might explain the failure you linked:
https://buildkite.com/julialang/julia-master/builds/24379#01886d90-17c9-48d9-9710-b2b18d515682

Reply:
`volatile` is only meaningful for special memory mapped hardware registers, and x86 doesn't have any of those

Reply:
I think what Jameson meant is that the notion of "volatile" is irrelevant here. The notion that is of import is "atomicity" (https://en.cppreference.com/w/c/language/atomic).
Either by marking the Bool `Atomic` or by protecting it with a lock. Before C11 volatile and atomic were often confounded.

From https://en.cppreference.com/w/c/language/volatile

> Note that volatile variables are not suitable for communication between threads; they do not offer atomicity, synchronization, or memory ordering. A read from a volatile variable that is modified by another thread without synchronization or concurrent modification from two unsynchronized threads is undefined behavior due to a data race.

Reply:
aha. Thank you both! That's clear now, then. 👍 Thanks for educating me!

So, a julia ReentrantLock is enough to also provide the LLVM / C concept of atomicity? We can guarantee that the updates to the value will be reflected across threads?

Now that we're discussing it, the answer seems to obviously be yes, otherwise how would anything work... OKAY thanks this is helpful.